### PR TITLE
Global transmission database integration

### PR DESCRIPTION
**Pull request by maartenbrinkerink**
PR to enable the integration of the Global Transmission Database (GTD) and other transmission related developments.

### Description

1. Pulls data from the Global Transmission Database (GTD)
2. Added functions for the formatting of GTD data/removed all connections to PLEXOS data for transmission
3. Added transmission centerpoints
4. Added functions to calculate transmission distance
5. Added functions to assign a distance dependent transmission technology per pathway
6. Added distance and technology dependent cost and efficiency parameters
7. Added transmission_parameters in config file to let users set costs/efficiency constants per transmission technology
8. Changes to output variables by integrating GTD data (e.g IAR/OAR, residual capacities, costs)
9. Changes to user_defined_capacity_transmission in config file to allow multiple user defined entries per transmission technology. Only changes to 'capacity' and 'build_year' are supported (i.e. to allow for step-wise residual capacities).

### Issue Ticket Number
#186, #120, #10, #172, #173, #46, #51

### CODE REVIEW

File: workflow/rules/retrieve.smk [hunk 1]
```diff
@@ -20,7 +22,13 @@ def get_external_links() -> dict[str,str]:
         "https://dataverse.harvard.edu/api/access/datafile/6040815",
 
         "ember_yearly_electricity_data.csv" :
-        "https://ember-climate.org/app/uploads/2022/07/yearly_full_release_long_format.csv"
+        "https://ember-climate.org/app/uploads/2022/07/yearly_full_release_long_format.csv",
```
Comment (trevorb1) [inline workflow/rules/retrieve.smk:25]: This has changed (must have been within the last few days)

https://storage.googleapis.com/emb-prod-bkt-publicdata/public-downloads/yearly_full_release_long_format.csv

Comment (trevorb1) [reply workflow/rules/retrieve.smk]: Corrected with [this commit ](https://github.com/OSeMOSYS/osemosys_global/pull/211/commits/922439c14506ade71cb2a9248f7e7fad9d86634a)

File: workflow/scripts/osemosys_global/transmission/main.py [hunk 2]
```diff
@@ -31,13 +49,18 @@
...
 def main(
-    plexos_prop: pd.DataFrame,
-    default_op_life: pd.DataFrame,
-    line_data: pd.DataFrame,
-    interface_data: pd.DataFrame,
+    default_op_life: dict[str, int],
+    gtd_exist : pd.DataFrame,
+    gtd_planned : pd.DataFrame,
+    gtd_mapping : dict[str, str],
+    centerpoints : list,
```
Comment (trevorb1) [inline workflow/scripts/osemosys_global/transmission/main.py:63]: `centerpoints` is not used in this function? Remove argument?

File: workflow/scripts/osemosys_global/transmission/data.py [hunk 3]
```diff
@@ -1,51 +1,180 @@
...
+    
+    df = df.rename(columns = cols)
+    
+    return df
 
-    # If from column has length 6 then it's the last three chars plus XX
-    df.loc[df["From"].str.len() == 6, "From"] = (df["From"].str[3:6] + "XX")
+def correct_gtd_data(df_exist, df_plan, region_mapping_dict, 
+                     CUSTOM_TRN_BA_DICT_FROM, CUSTOM_TRN_BA_DICT_TO,
+                     CUSTOM_TRN_BA_MISSING):
```
Comment (trevorb1) [inline workflow/scripts/osemosys_global/transmission/data.py:45]: Not a required change; just a FYI. 

According to PEP8, [constants should be defined as all caps](https://peps.python.org/pep-0008/#constants) (as you did for the last three arguments in the constants file), but [function arguments and variables should still follow snakecase formatting](https://peps.python.org/pep-0008/#function-and-variable-names).

File: workflow/scripts/osemosys_global/transmission/data.py [hunk 4]
```diff
@@ -1,51 +1,180 @@
...
+    within a single GTD region (e.g. USASN and USASS in MISO) are not automatically 
+    pulled from the GTD datasets so need to be manually added. Furthermore, pathways
+    within the GTD datasets that are not selected as the pathways to which residual
+    capacity should be allocated (defined in the CUSTOM_TRN_BA_DICT_FROM and
+    CUSTOM_TRN_BA_DICT_TO constants) also need to be manually added."""
+    df_missing = pd.DataFrame(CUSTOM_TRN_BA_MISSING, columns = ['TECHNOLOGY'])
+    df_missing['From'] = df_missing['TECHNOLOGY'].str[3:8]
+    df_missing['To'] = df_missing['TECHNOLOGY'].str[8:13]
+    df_missing['YEAR'] = '-'
+    df_missing['VALUE'] = 0
```
Comment (trevorb1) [inline workflow/scripts/osemosys_global/transmission/data.py:109]: Does PLEXOS-World regions follow [NARIS/ReEDS boundaries](https://www.nrel.gov/analysis/naris.html)? If so, we can just grab transmission data from there and not worry about doing this custom mapping.

Comment (trevorb1) [reply workflow/scripts/osemosys_global/transmission/data.py:109]: [Figure 4](https://www.nrel.gov/docs/fy21osti/78195.pdf) shows spatial structure

File: workflow/scripts/osemosys_global/transmission/main.py [hunk 5]
```diff
@@ -118,53 +164,57 @@ def main(
     
     # assign capacity to activity unit to transmission + distribution techs
     cap_activity_trn = create_trn_dist_capacity_activity(iar_trn, oar_trn)
-    cap_activity = pd.concat([capact_base, cap_activity_trn]).drop_duplicates(subset=["REGION", "TECHNOLOGY"], keep="last")
+    cap_activity = pd.concat([capact_base, cap_activity_trn]
+                             ).drop_duplicates(subset=["REGION", "TECHNOLOGY"], keep="last")
     
     # OUTPUT CSV's
```
Comment (trevorb1) [inline workflow/scripts/osemosys_global/transmission/main.py:170]: Should we be writing out only the `TRN` techs, then having a rule to concat all the different CSVs together. Like the `powerplant/CapitalCosts.csv` only contains power plant capital costs, and the `transmission/CapitalCosts.csv` only contains transmission capital costs. 

That way rules have less dependencies?

Comment (trevorb1) [reply workflow/scripts/osemosys_global/transmission/main.py:170]: Also mean you dont need to read in data like capacity to activity units for powerplant technologies

File: workflow/scripts/osemosys_global/transmission/main.py [hunk 6]
```diff
@@ -53,57 +76,80 @@ def main(
...
     
-    # Set capital and fixed transmission costs.
-    df_cap_cost_trn_final, df_fix_cost_trn_final = get_transmission_costs(line_data, oar_trn,
-                                                                    cap_cost_base, fix_cost_base,
-                                                                    start_year, end_year,
-                                                                    region_name)
-
     # Set operational life for transmission.
-    df_op_life_trn_final = set_op_life_transmission(iar_trn, oar_trn, 
+    op_life_trn = set_op_life_transmission(iar_trn, oar_trn,
```
Comment (trevorb1) [inline workflow/scripts/osemosys_global/transmission/main.py:110]: Does this function require both `iar_trn` and `oar_trn`? What transmission techs exist in one but not the other?

File: config/config.yaml [hunk 7]
```diff
@@ -50,23 +50,43 @@ solver: "cbc" # cbc, cplex, gurobi
...
+  #         capex_cost,
+  #         annual_fixed_O&M_cost,
+  #         efficiency]
+  trn1: [TRNINDEAINDNE, 5, 1975, 2030, 10, 861, 30, 95]
+  trn2: [TRNINDEAINDNE, 1, 1990, 2030, 10, 861, 30, 95]
+  trn3: [TRNINDEAINDNE, 2, 2035, 2030, 10, 861, 30, 95]
+  trn4: [TRNINDNOINDSO, 0, 2020, 2025, 0.5, 900, 32, 92]
+  # Per technology, only the capacity and build_year
+  # parameters can deviate from each other. For all other
+  # parameters only the last entry will be used.
```
Comment (trevorb1) [inline config/config.yaml:76]: Why the change to defining `trn1`, `trn2`, ect... ?

File: workflow/scripts/osemosys_global/transmission/main.py [hunk 8]
```diff
@@ -53,57 +76,80 @@ def main(
     
     # CALL FUNCTIONS
     
+    # Correct the input data from the Global Transmission Database (GTD).
+    gtd_exist_corrected, gtd_planned_corrected = correct_gtd_data(gtd_exist,
```
Comment (trevorb1) [inline workflow/scripts/osemosys_global/transmission/main.py:80]: Is it easier to just return one dataframe? Or just move up the res capacity calc right under this to just have one transmission datastructure to pass around?

**Review by trevorb1 — COMMENTED** [review]
I don't see any breaking changes! Please see my general comments.

File: workflow/scripts/osemosys_global/transmission/data.py (hunk 3 above)
Comment (maartenbrinkerink) [reply workflow/scripts/osemosys_global/transmission/data.py:45]: Fixed for both powerplant and transmission scripts

File: workflow/scripts/osemosys_global/transmission/data.py (hunk 4 above)
Comment (maartenbrinkerink) [reply workflow/scripts/osemosys_global/transmission/data.py:109]: No it mostly follows NEMS see Figure 3 [https://www.eia.gov/outlooks/aeo/nems/documentation/electricity/pdf/EMM_2022.pdf](url). I think that when we start incorporating GEM data later this year/early next year we should reconsider the US nodes within OG. I.e. if we have shapefiles/boundaries we can allocate individual plants within GEM more accurately.

File: workflow/scripts/osemosys_global/transmission/main.py (hunk 6 above)
Comment (maartenbrinkerink) [reply workflow/scripts/osemosys_global/transmission/main.py:110]: No it doesn't! Fixed

File: workflow/scripts/osemosys_global/transmission/main.py (hunk 8 above)
Comment (maartenbrinkerink) [reply workflow/scripts/osemosys_global/transmission/main.py:80]: Some of the functions that use `gtd_exist_corrected` and `gtd_planned_corrected` require it in the format as initially generated (e.g. with the from and to columns). The `res_capacity_transmission` output format is sligthly different. No biggy to change if necessary but the transmission scripts are quite quick as it is already so might not be worth the effort to reduce the nr. of dfs that are being passed around. I also find it convenient to have the existing/planned df's separately in case they need to be called in future functions. Let me know!

File: config/config.yaml (hunk 7 above)
Comment (maartenbrinkerink) [reply config/config.yaml:76]: The entries are read into the different scripts as dictionaries. Keeping the `technology` name as key would lead to the different residual capacity entries to be read incorrectly. @trevorb1 If you have a better alternative please do let me know (I didn't love setting it up this way either but ran out of brain space... ).

Comment (trevorb1) [reply config/config.yaml:76]: Ahhh fair enough; good point! I think until we come up with a better method to do user defined capacity, this works just great. :)

File: workflow/scripts/osemosys_global/transmission/main.py (hunk 8 above)
Comment (trevorb1) [reply workflow/scripts/osemosys_global/transmission/main.py:80]: Fair point and your right; not urgent and it works as is :)

**Review by trevorb1 — APPROVED** [review]
